### PR TITLE
fix(completions): resolve OpenAI bare model aliases (gpt-4.1) in the public completions API

### LINKED ISSUE #2745: fix(completions): gpt-4.1 fails with "Failed to create LLM backend for model: gpt-4.1"
## What

`POST /api/ai/v1/completions` with `model: "gpt-4.1"` returns HTTP 200 and an SSE error event:

```
{"type":"error","message":"Failed to create LLM backend for model: gpt-4.1"}
```

`gpt-4o` and `gpt-5.4` work on the same request. Verified live 2026-09-12 with a multimodal message and a `json_schema` response format; the failure is independent of both (plain text request fails the same way).

## Where to look

The backend factory used by `apps/client/server/chatCompletion/external/sseRoute.ts` and the OpenAI model registry in `b4m-core/llm-adapters/src/openaiBackend.ts` (~670-754 for vision-capable entries). Either the id is missing from the registry or the factory's model-to-provider mapping does not cover it.

## Expected

Either `gpt-4.1` (and `gpt-4.1-mini`) resolve to the OpenAI backend, or the model is removed from any list that advertises it.

Found while building the 4XCiv icon judge (MillionOnMars/4xciv-engine#367).

### PR DESCRIPTION
**Pull request by wescarda**
# Pull Request

## Description

`POST /api/ai/v1/completions` with `model: "gpt-4.1"` returned an SSE error event, "Failed to create LLM backend for model: gpt-4.1", instead of a completion. `gpt-4o` and `gpt-5.4` worked on the same request.

Root cause: our model catalog stores GPT-4.1's canonical id as OpenAI's dated snapshot (`ChatModels.GPT4_1 = 'gpt-4.1-2025-04-14'` in `b4m-core/common/src/models.ts`), not the bare `gpt-4.1` string. OpenAI's own API accepts `gpt-4.1` as an alias for whatever snapshot is current, and a public API caller using that convention hit a plain `.find(m => m.id === model)` in the shared `executeCompletion` path (`b4m-core/services/src/cliCompletions.ts`), which never matches the dated id. `gpt-4o` and `gpt-5.4` happened to work because those two enum values are already bare.

The same "dated-id-only" pattern also affects a handful of other OpenAI models (`o1`, `o3`, `o1-preview`, `o1-mini`, `o3-mini`, `o4-mini`, `gpt-4.5-preview`), but this PR only closes the aliases the issue reports (`gpt-4.1`, `gpt-4.1-mini`, `gpt-4.1-nano`) to keep the change targeted.

Closes https://github.com/Bike4Mind/bike4mind/issues/2745

## Changes

- `cliCompletions.ts`: resolve the known bare OpenAI aliases (`gpt-4.1`, `gpt-4.1-mini`, `gpt-4.1-nano`) to their catalog dated id once, up front in `executeCompletion`, so every downstream use of `model` (backend lookup, cost/credit calc, the completion call itself, logged usage) sees one consistent id. Canonical dated ids are left unchanged, so any session/agent already pinned to one is unaffected.
- Alias lookup uses a `Map`, not a plain object, since the key is caller-controlled input (a plain-object lookup would return inherited `Object.prototype` members for keys like `"constructor"`).
- Added `cliCompletions.modelAlias.test.ts`: covers the bare alias resolving correctly, the canonical dated id still working directly, an unknown model still raising the original error, and the prototype-key edge case.

## Guide for Testers

Backend-only fix in the shared completions path; no UI surface to click through.

1. Run `pnpm --filter @bike4mind/services exec vitest run src/cliCompletions.modelAlias.test.ts` — all 4 tests pass.
2. Functionally: call `POST /api/ai/v1/completions` with `{"model": "gpt-4.1", "messages": [...]}` and a valid API key. Expect a normal streamed completion instead of the `Failed to create LLM backend for model: gpt-4.1` SSE error event. Same for `gpt-4.1-mini` and `gpt-4.1-nano`.
3. Regression: the same request with `model: "gpt-4o"` or `model: "gpt-5.4"` should continue to work unchanged.

## Additional Information

Out of scope: `o1`, `o3`, `o1-preview`, `o1-mini`, `o3-mini`, `o4-mini`, and `gpt-4.5-preview` share the same dated-id-only catalog pattern and would fail the same way for a bare-alias request. Not fixed here since issue #2745 only reports the gpt-4.1 family.

## Developer Notes (Optional)

None beyond the alias map itself, which is intentionally narrow (see comment in `cliCompletions.ts`) rather than a generic mechanism, since no such mechanism already existed and this fix should not be the place to introduce one.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] N/A: AdminSettings / UI/UX / docs / telemetry checklist items (backend-only, no new settings, UI, or telemetry fields)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: b4m-core/services/src/cliCompletions.ts [hunk 1]
```diff
@@ -147,12 +147,30 @@ function estimateInputTokens(messages: IMessage[]): number {
...
+]);
+
 /**
  * Shared LLM completion logic
  * Used by Next.js API route, Lambda function, and available for 3rd party integrations
  */
 export async function executeCompletion(params: CompletionParams): Promise<void> {
-  const { userId, model, messages, options, db, logger, onChunk, apiKeyInfo } = params;
+  const { userId, messages, options, db, logger, onChunk, apiKeyInfo } = params;
+  const model = OPENAI_BARE_MODEL_ALIASES.get(params.model) ?? params.model;
```
Comment (onoya) [inline b4m-core/services/src/cliCompletions.ts]: Nice fix for the actual completion path — but tracing `model` to its other terminal sink: `apps/client/server/utils/logCompletionAnalytics.ts:79` does its own `models.find(m => m.id === (body.model as ChatModels))` against the *original, unresolved* request body, independently of this resolution. All three callers that hit this bug (`sseRoute.ts`, `wsRoute.ts`, `cliCompletion.ts`) call `logCompletionAnalytics` right after `executeCompletion`, with the same raw `model` string.

Before this PR, `executeCompletion` threw before that call ever ran for `gpt-4.1`, so it was dead code for these models. After this PR, the completion succeeds and `logCompletionAnalytics` fails its own lookup every time — logging `'Model not found for credits calculation'` and reporting `creditsUsed: 0` in the `COMPLETION_API_COMPLETED` analytics event, even though the user was correctly charged a nonzero amount.

Simplest fix: export this resolution (the map, or a `resolveOpenAiBareModelAlias()` helper) and apply it in `logCompletionAnalytics.ts` before its `.find()` too, so the two lookups can't drift again.

**Review by onoya — CHANGES_REQUESTED** [review]
This resolves the real bug — `gpt-4.1`/`-mini`/`-nano` bare aliases now resolve to the catalog's dated snapshot id at the single choke point shared by every completion entry point, and the switch to a `Map` for a caller-controlled lookup key is exactly right (nice touch testing the `constructor` case too). The approach is sound and the scope you drew around `o1`/`o3`/etc. is reasonable.

Left one inline: tracing `model` past `cliCompletions.ts` to its other terminal sink, `logCompletionAnalytics.ts` turns out to run its own unaliased lookup — so it'll silently misreport `creditsUsed: 0` for exactly the traffic this PR unblocks. That needs to close before merge.

Non-blocking, happy to file as issues if you want them tracked:
- The same dated-id-only pattern still bites `o1`, `o3`, `o1-preview`, `o1-mini`, `o3-mini`, `o4-mini`, `gpt-4.5-preview` — you've scoped that out explicitly and that's a fine call, just flagging it's still out there.
- Once the analytics fix lands, consider exporting `OPENAI_BARE_MODEL_ALIASES` rather than growing a second hand-synced copy of it.

File: b4m-core/services/src/cliCompletions.ts [hunk 2]
```diff
@@ -147,12 +147,39 @@ function estimateInputTokens(messages: IMessage[]): number {
...
+ * current, but our catalog stores the dated snapshot id itself (ChatModels.GPT4_1 etc. in
+ * @bike4mind/common), so a caller using OpenAI's convention otherwise finds no catalog
+ * entry. Resolved once, up front, so every downstream use of `model` (backend lookup,
+ * cost/credit calc, the completion call itself, logged usage) sees one consistent id.
+ *
+ * A Map, not a plain object: `model` is caller-controlled, and a plain object lookup
+ * keyed by an arbitrary string returns inherited properties for keys like "constructor"
+ * or "toString" instead of undefined.
+ */
+const OPENAI_BARE_MODEL_ALIASES: ReadonlyMap<string, ChatModels> = new Map([
```
Comment (wescarda) [inline b4m-core/services/src/cliCompletions.ts:161]: **Second-pass review note (not applied):** this map duplicates the caller-id-resolution pattern already established by `resolveDeprecatedModelId`/`DEPRECATED_MODEL_MAP` in `b4m-core/llm-adapters/src/resolveDeprecatedModel.ts`, which `llm-adapters/src/index.ts` documents as where model-id resolution belongs -- so the codebase now has two independent tables that normalize a caller-supplied model id.

Considered folding `gpt-4.1`/`gpt-4.1-mini`/`gpt-4.1-nano` into `DEPRECATED_MODEL_MAP` instead, but that map's own doc comment and enforced test (`deprecationDate <-> mapping` invariant) frame it strictly as a *deprecated-to-successor* resolver -- these models are fully current, not deprecated, so adding them there would blur that invariant and is also what `buildSupersededIndex` uses to tell users a pinned model was replaced (a wrong signal for a live model). Kept this as a small, separately-scoped alias table instead, now exported (`resolveOpenAiBareModelAlias`) so at least the two call sites inside this PR (`executeCompletion` and `logCompletionAnalytics`) share one instance rather than drifting.

Not blocking -- flagging so the two-tables tension is visible in review rather than only in my own reasoning.

File: b4m-core/services/src/cliCompletions.ts (hunk 1 above)
Comment (wescarda) [reply b4m-core/services/src/cliCompletions.ts]: Fixed in b4f06662975af133e30cfe2cfb3e4016f68c9c3a (commit 'fix(completions): resolve the same bare alias in completion analytics') -- exported `resolveOpenAiBareModelAlias()` from `cliCompletions.ts` and applied it in `logCompletionAnalytics.ts:82` before its own `.find()`, so both call sites resolve the same way instead of drifting. Added a regression test (`logCompletionAnalytics.test.ts`) asserting a bare `gpt-4.1` request no longer logs the spurious "model not found" error and reports the correct nonzero `creditsUsed`, plus a companion test that a genuinely unknown model still logs it. Thanks for catching this -- exact same bug and fix shape I landed independently in the same pass.

**Review by onoya — APPROVED** [review]
LGTM! 🚀

Followed `model` all the way through `executeCompletion` to `llm.complete()`, `subtractCredits`, and now `logCompletionAnalytics.ts` too — the second sink is fixed and tested, and there's no third one hiding in `embedRoute.ts` (`agent.preferredModel` is Zod-enum-validated to a canonical id, so the bare-alias problem can't actually reach it). Good instinct fixing the analytics gap independently of the review thread, and the `constructor`-key test is a genuinely good one — it proves the prototype-pollution class of bug is prevented, not just avoided by convention.

A couple of things worth a follow-up rather than a blocker, happy to file issues if you want them tracked:
- Two independent model-id-resolution tables now exist (`OPENAI_BARE_MODEL_ALIASES` in `cliCompletions.ts` vs. `DEPRECATED_MODEL_MAP` in `llm-adapters`). You already reasoned through why they shouldn't merge (thread on `cliCompletions.ts:161`) — just worth a one-line pointer comment on `DEPRECATED_MODEL_MAP` so the next person doesn't rediscover the split.
- The PR body names seven other dated-id-only models (`o1`, `o3`, `o1-preview`, `o1-mini`, `o3-mini`, `o4-mini`, `gpt-4.5-preview`) sharing this pattern — worth a tracking issue so that list doesn't only live in prose.

Nit, take or leave: `OPENAI_BARE_MODEL_ALIASES.get()` is exact-match, so a caller sending `"GPT-4.1"` (uppercase) still 404s the old way. Unlikely given OpenAI's own convention is lowercase, but a `.toLowerCase()` on the lookup key would close it for free if you're back in here.

File: b4m-core/services/src/cliCompletions.ts (hunk 1 above)
Comment (onoya) [reply b4m-core/services/src/cliCompletions.ts]: Verified — logCompletionAnalytics.ts:79-82 now calls resolveOpenAiBareModelAlias(body.model) before its own .find(), and logCompletionAnalytics.test.ts pins the nonzero-credits/no-spurious-error behavior for a bare alias. Resolving; reopen if I've read this wrong.

File: b4m-core/services/src/cliCompletions.ts (hunk 2 above)
Comment (onoya) [reply b4m-core/services/src/cliCompletions.ts:161]: Verified — kept as a separate table and exported (resolveOpenAiBareModelAlias, cliCompletions.ts:161-173) rather than folded into DEPRECATED_MODEL_MAP, matching the reasoning here. Left a note in the approval body suggesting a one-line pointer comment on DEPRECATED_MODEL_MAP for the next reader. Resolving; reopen if you'd rather leave this open for that.
